src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,936 @@
+
+        :return: A BlkDiagMatrix like self.
+        """
+
+        C = BlkDiagMatrix(self.partition, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = np.abs(self[i])
+
+        return C
+
+    def __abs__(self):
+        """
+        Operator overload for absolute value of BlkDiagMatrix instance.
+        """
+
+        return self.abs()
+
+    def pow(self, val, inplace=False):
+        """
+        Compute the elementwise power of BlkDiagMatrix instance.
+        :param inplace: Boolean, when set to True change values in place,
+        otherwise return a new instance (default).
+        :return: A BlkDiagMatrix like self.
+        """
+
+        if inplace:
+            for i in range(self.nblocks):
+                self[i] **= val
+            C = self
+        else:
+            C = BlkDiagMatrix(self.partition, dtype=self.dtype)
+            for i in range(self.nblocks):
+                C[i] = np.power(self[i], val)
+        return C
+
+    def __pow__(self, val):
+        """
+        Operator overload for inplace pow of BlkDiagMatrix instance.
+        """
+
+        return self.pow(val)
+
+    def __ipow__(self, val):
+        """
+        Compute the in-place elementwise power of BlkDiagMatrix instance.
+
+        :return: self raised to power, elementwise.
+        """
+
+        return self.pow(val, inplace=True)
+
+    def norm(self, order=2):
+        """
+        Compute the norm of a BlkDiagMatrix instance.
+
+        :param order: Norm order, see np.norm. Defaults to order 2 norm.
+        :param inplace: Boolean, when set to True change values in place,
+        otherwise return a new instance (default).
+        :return: The norm of the BlkDiagMatrix instance.
+        """
+
+        implemented_orders = (2,)
+        if order not in implemented_orders:
+            raise NotImplementedError(
+                "Order {} not yet implemented, only {}".format(
+                    ord, implemented_orders))
+
+        return np.max([norm(blk, ord=order) for blk in self])
+
+    def transpose(self):
+        """
+        Get the transpose matrix of a BlkDiagMatrix instance.
+
+        :return: The corresponding transpose form as a BlkDiagMatrix.
+        """
+
+        T = BlkDiagMatrix(self.partition, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            T[i] = self[i].T
+
+        return T
+
+    @property
+    def T(self):
+        """
+        Syntactic sugar for self.transpose().
+        """
+
+        return self.transpose()
+
+    def dense(self):
+        """
+        Convert list representation of BlkDiagMatrix instance into full matrix.
+
+        :param blk_diag: The BlkDiagMatrix instance.
+        :return: The BlkDiagMatrix instance including the zero elements of
+        non-diagonal blocks.
+        """
+
+        return block_diag(self.data)
+
+    @staticmethod
+    def from_list(blk_diag, dtype=np.float64):
```

Comment:
Shouldn't `dtype` be inferred from the first argument (also below)?

Reply:
Not necessarily, I don't assume it is a numpy array instance, just something that can become one.  Usually it is, but for generality I thought it might be better to admit a list of 2d-lists...